### PR TITLE
SONARPY-208 Rule: Unused local variables should be removed

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.python.PythonCheck;
+import org.sonar.python.api.PythonGrammar;
+import org.sonar.python.semantic.Symbol;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+
+@Rule(
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java]: MINOR Reorder annotation arguments to match the order of declaration. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3340)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [semantic](https://github.com/SonarSource/sonar-python/blob/3ceb2731d62e1ec275f7c236dafde9fa5a81c175/python-squid/src/main/java/org/sonar/python/semantic): Add a 'package-info.java' file to document the 'semantic' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.semantic;
+
+import com.google.common.collect.ImmutableSet;
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.AstNodeType;
+
+import java.util.*;
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java]: MAJOR Explicitly import the specific classes needed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2208)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [semantic](https://github.com/SonarSource/sonar-python/blob/698bda1a17c3c062a5cb0607859f89761a1ecc11/python-squid/src/main/java/org/sonar/python/semantic): Add a 'package-info.java' file to document the 'semantic' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  key = "S1481",
+  name = "Unused local variables should be removed",
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java]: MINOR Reorder annotation arguments to match the order of declaration. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3340)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        PythonGrammar.FUNCDEF,
+        PythonGrammar.CLASSDEF,
+        PythonGrammar.ATOM,
+        PythonGrammar.DOTTED_NAME);
+    }
+
+    @Override
+    public void visitNode(AstNode node) {
+      super.visitNode(node);
+      if (node.is(PythonGrammar.ATOM, PythonGrammar.DOTTED_NAME) && currentScopeRootTree().is(PythonGrammar.CLASSDEF)) {
```
Comment (ivandalbosco) [inline python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java]: @pynicolas If I remove the fragment
`currentScopeRootTree().is(PythonGrammar.CLASSDEF)`
from this line, the unit test still works. So the code fragment above can be removed, as this class is private and only used in a "class scope" context.

File: python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  @Override
+  public void visitNode(AstNode functionTree) {
+    for (Symbol symbol : getContext().symbolTable().symbols(functionTree)) {
+      checkSymbol(symbol);
+    }
+  }
+
+  private void checkSymbol(Symbol symbol) {
```
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java]: @pynicolas We come up with a FN on x with the following code:
```
    x = 3
    f = lambda x : x**2

```

File: python-squid/src/main/java/org/sonar/python/PythonVisitorContext.java [hunk 6]
```diff
@@ -21,15 +21,20 @@
...
 public class PythonVisitorContext {
 
   private final AstNode rootTree;
   private final PythonFile pythonFile;
   private final RecognitionException parsingException;
+  private SymbolTableBuilderVisitor symbolTable = null;
 
   public PythonVisitorContext(AstNode rootTree, PythonFile pythonFile) {
     this(rootTree, pythonFile, null);
+    symbolTable = new SymbolTableBuilderVisitor();
```
Comment (pynicolas) [inline python-squid/src/main/java/org/sonar/python/PythonVisitorContext.java]: @ivandalbosco Changed in f86f490

File: python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+
+    private Scope currentScope() {
+      return scopesByRootTree.get(currentScopeRootTree());
+    }
+
+  }
+
+  @Override
+  public Set<Symbol> symbols(AstNode scopeTree) {
+    return scopesByRootTree.get(scopeTree).symbols();
```
Comment (pynicolas) [inline python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java]: @ivandalbosco Fixed in cf1d221

File: python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    new SecondPhaseVisitor().scanFile(context);
+  }
+
+  @Override
+  public void visitFile(AstNode node) {
+    scopesByRootTree = new HashMap<>();
+    allReadUsages = new HashSet<>();
+  }
+
+  private class ScopeVisitor extends PythonVisitor {
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java]: MAJOR Make this a "static" inner class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java (hunk 5 above)
Comment (pynicolas) [inline python-checks/src/main/java/org/sonar/python/checks/UnusedLocalVariableCheck.java]: @ivandalbosco Ticket created: https://jira.sonarsource.com/browse/SONARPY-210

File: python-squid/src/test/resources/semantic/symbols.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
+a = 1
+a = 2
+b = 1
+
```
Comment (pynicolas) [inline python-squid/src/test/resources/semantic/symbols.py]: @ivandalbosco Ticket created: https://jira.sonarsource.com/browse/SONARPY-211

**Pull request merged**